content/binary-api.md

```diff
@@ -0,0 +1,227 @@
+The annotated definition will get a public accessor.
+
+This can be used to access `private`/`private[this]` definitions within inline definitions.
+
+Example:
+~~~ scala
+class C {
+  @binaryAPIAccessor private def privateAPI: Int = ...
+  @binaryAPIAccessor def publicAPI: Int = ...
+}
+~~~
+will generate the following bytecode signatures
+~~~ java
+public class C {
+  public C();
+  private int privateAPI();
+  public int publicAPI();
+  public final int C$$inline$privateAPI();
+  public final int C$$inline$publicAPI();
+}
+~~~
+
+Note that the change from `private[this]` to package private, protected or public is a binary compatible change.
+Removing this annotation is a binary incompatible change.
+
+In the bytecode, `@binaryAPIAccessor` generated accessors will have the [ACC_PUBLIC | ACC_SYNTHETIC](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.1-200-E.1) flag.
+
+#### Binary API and inlining
+
+A non-public reference in an inline method is handled as follows:
+ - if the reference is a `@binaryAPI` the reference is used;
+ - if the reference is a `@binaryAPIAccessor` the accessor is used;
+ - otherwise, an accessor is automatically generated and used. We will emit a warning with an actionable diagnostic containing the code needed to migrate to `@binaryAPI` or `@binaryAPIAccessor`.
+
+Example 3:
+~~~ scala
+class C {
+  @binaryAPIAccessor private def a: Int = ...
+  private def b: Int = ...
+  @binaryAPI protected def c: Int = ...
+  protected def d: Int = ...
+  inline def foo: Int = a + b + c + d
+}
+~~~
+before inlining the compiler will generate the accessors for inlined definitions
+~~~ scala
+class C {
+  @binaryAPIAccessor private def a: Int = ...
+  private def b: Int = ...
+  @binaryAPI protected def c: Int = ...
+  protected def d: Int = ...
+  final def C$inline$a: Int = ... // generated by `@binaryAPIAccessor`
+  final def C$inline$b: Int = ... // warn: `b` should be annotated with `@binaryAPIAccessor` + migration code
+  final def C$inline$d: Int = ... // warn: `d` should be annotated with `@binaryAPI` + migration code
+  inline def foo: Int = C$inline$a + C$inline$b + c + C$inline$d
+}
+~~~
+
+### Specification
+
+We must add `binaryAPI` and `binaryAPIAccessor` to the standard library.
+
+```scala
+package scala.annotation
+
+final class binaryAPI extends scala.annotation.StaticAnnotation
+final class binaryAPIAccessor extends scala.annotation.StaticAnnotation
+```
+
+#### `@binaryAPI` annotation
+
+* Only valid on `def`, `val`, `lazy val`, `var`, `object`, and `given`.
+* TASTy will contain references to non-public definitions that are out of scope but `@binaryAPI`. TASTy already allows those references.
+* Annotated definition will be public in the generated bytecode. Definitions should be made public as early as possible in the compiler phases, as this can remove the need to create other accessors. It should be done after we check the accessibility of references.
```

Comment:
```suggestion
* The annotated definitions will be public in the generated bytecode. Definitions should be made public as early as possible in the compiler phases, as this can remove the need to create other accessors. It should be done after we check the accessibility of references.
```

Reply:
About the second part “Definitions should be made public as early as possible in the compiler phases, as this can remove the need to create other accessors”: is that an implementation detail? Or is it part of the specification? If that’s part of the specification, could you please explain which accessor will avoid being generated?